components/tikv_util/src/resizable_threadpool.rs

```diff
@@ -1,161 +1,350 @@
+
         ResizableRuntime {
-            size: 0,
+            size: thread_size,
+            count: 0,
             thread_name: thread_name.to_owned(),
-            pool: None,
+            gc_runtime: DeamonRuntime {
+                inner: Some(keeper),
+                tracker: TaskTracker::new(),
+            },
+            current_runtime: Arc::new(Mutex::new(DeamonRuntime {
+                inner: Some(new_runtime),
+                tracker: TaskTracker::new(),
+            })),
             replace_pool_rule,
             after_adjust,
         }
     }
 
+    pub fn size(&self) -> usize {
+        self.size
+    }
+
     pub fn spawn<Fut>(&self, fut: Fut)
     where
         Fut: Future<Output = ()> + Send + 'static,
     {
-        self.pool
-            .as_ref()
-            .expect("ResizableRuntime: please call adjust_with() before spawn()")
-            .spawn(fut);
+        let handle = self.handle();
+        handle.spawn(fut);
     }
 
-    /// Lazily adjust the thread pool's size
-    pub fn adjust_with(&mut self, new_size: usize) {
-        if self.size == new_size {
-            return;
-        }
-        // TODO: after tokio supports adjusting thread pool size(https://github.com/tokio-rs/tokio/issues/3329),
-        //   adapt it.
-        let pool = (self.replace_pool_rule)(new_size, &self.thread_name)
-            .expect("failed to create tokio runtime for backup worker.");
-        self.pool = Some(DaemonRuntime::from_runtime(pool));
-        self.size = new_size;
-        (self.after_adjust)(new_size);
-    }
-
-    pub fn block_on<F>(&self, f: F) -> F::Output
+    pub fn block_on<Fut>(&self, fut: Fut)
     where
-        F: Future,
+        Fut: Future<Output = ()> + Send + 'static,
     {
-        self.pool
-            .as_ref()
-            .expect("ResizableRuntime: please call adjust_with() before block_on()")
-            .0
-            .as_ref()
-            .unwrap()
-            .block_on(f)
+        let handle = self.handle();
+        handle.block_on(fut);
     }
-}
-
-#[derive(Clone)]
-pub struct ResizableRuntimeHandle {
-    inner: Arc<RwLock<ResizableRuntime>>,
-}
 
-impl ResizableRuntimeHandle {
-    pub fn new(runtime: ResizableRuntime) -> Self {
-        ResizableRuntimeHandle {
-            inner: Arc::new(RwLock::new(runtime)),
+    pub fn handle(&self) -> DeamonRuntimeHandle {
+        DeamonRuntimeHandle {
+            inner: Arc::downgrade(&self.current_runtime),
         }
     }
 
-    pub fn spawn<Fut>(&self, fut: Fut)
-    where
-        Fut: Future<Output = ()> + Send + 'static,
-    {
-        let inner = self.inner.read().unwrap();
-        inner.spawn(fut);
-    }
+    // TODO: after tokio supports adjusting thread pool size(https://github.com/tokio-rs/tokio/issues/3329),
+    //   adapt it.
+    pub fn adjust_with(&mut self, new_size: usize) {
+        let mut runtime_guard = self.current_runtime.lock().unwrap();
```

Comment:
The `runtime_guard`'s scope should not exceed that beyond the operation of `std::mem::replace`. Once `replace` is done the guard should be dropped.